packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java

```diff
@@ -0,0 +1,311 @@
+   * {@link DartMessenger}. This latest value can also be retrieved through the {@link
+   * #getMediaOrientation()} accessor.
+   *
+   * <p>If the device's ACCELEROMETER_ROTATION setting is enabled the {@link
+   * DeviceOrientationManager} will report orientation updates based on the sensor information. If
+   * the ACCELEROMETER_ROTATION is disabled the {@link DeviceOrientationManager} will fallback to
+   * the deliver orientation updates based on the UI orientation.
+   */
+  public void start() {
+    startSensorListener();
+    startUIListener();
+  }
+
+  /** Stops listening for orientation updates. */
+  public void stop() {
+    stopSensorListener();
+    stopUIListener();
+  }
+
+  /**
+   * Returns the last captured orientation in degrees based on sensor or UI information.
+   *
+   * <p>The orientation is returned in degrees and could be one of the following values:
+   *
+   * <ul>
+   *   <li>0: Indicates the device is currently in portrait.
+   *   <li>90: Indicates the device is currently in landscape left.
+   *   <li>180: Indicates the device is currently in portrait down.
+   *   <li>270: Indicates the device is currently in landscape right.
+   * </ul>
+   *
+   * @return The last captured orientation in degrees
+   */
+  public int getMediaOrientation() {
+    return this.getMediaOrientation(this.lastOrientation);
+  }
+
+  /**
+   * Returns the device's orientation in degrees based on the supplied {@link
+   * PlatformChannel.DeviceOrientation} value.
+   *
+   * <p>
+   *
+   * <ul>
+   *   <li>PORTRAIT_UP: converts to 0 degrees.
+   *   <li>LANDSCAPE_LEFT: converts to 90 degrees.
+   *   <li>PORTRAIT_DOWN: converts to 180 degrees.
+   *   <li>LANDSCAPE_RIGHT: converts to 270 degrees.
+   * </ul>
+   *
+   * @param orientation The {@link PlatformChannel.DeviceOrientation} value that is to be converted
+   *     into degrees.
+   * @return The device's orientation in degrees.
+   */
+  public int getMediaOrientation(PlatformChannel.DeviceOrientation orientation) {
+    int angle = 0;
+
+    // Fallback to device orientation when the orientation value is null
+    if (orientation == null) {
+      orientation = getUIOrientation();
+    }
+
+    switch (orientation) {
+      case PORTRAIT_UP:
+        angle = 0;
+        break;
+      case PORTRAIT_DOWN:
+        angle = 180;
+        break;
+      case LANDSCAPE_LEFT:
+        angle = 90;
+        break;
+      case LANDSCAPE_RIGHT:
+        angle = 270;
+        break;
+    }
+
+    if (isFrontFacing) {
+      angle *= -1;
+    }
+
+    return (angle + sensorOrientation + 360) % 360;
+  }
+
+  private void startSensorListener() {
+    if (orientationEventListener != null) {
+      return;
+    }
+    orientationEventListener =
+        new OrientationEventListener(activity, SensorManager.SENSOR_DELAY_NORMAL) {
+          @Override
+          public void onOrientationChanged(int angle) {
+            if (!isSystemAutoRotationLocked()) {
```

Comment:
The logic in these handlers isn't getting tested (e.g., if one of these checks were reversed, no tests would catch it). If the testing is going to be via calling private methods rather than mocking the listeners—I'm not sure how feasible that is; I've done very little Java—then as much logic as possible should be in those test-called private methods.